src/schedule/zephyr_domain.c

```diff
@@ -177,84 +153,69 @@ static int zephyr_domain_unregister(struct ll_schedule_domain *domain,
 	if (num_tasks)
 		return 0;
 
-	k_work_cancel_delayable(&zdata[core].work);
+	k_timer_stop(&zephyr_domain->timer);
+	zephyr_domain->domain_thread[core].handler = NULL;
+
+	k_thread_abort(&zephyr_domain->domain_thread[core].ll_thread);
 
 	tr_info(&ll_tr, "zephyr_domain_unregister domain->type %d domain->clk %d",
 		domain->type, domain->clk);
 
 	return 0;
 }
 
+/* Can be called on any core, sets timer IRQ for core 0 if CONFIG_SMP_BOOT_DELAY */
 static void zephyr_domain_set(struct ll_schedule_domain *domain, uint64_t start)
 {
 	struct zephyr_domain *zephyr_domain = ll_sch_domain_get_pdata(domain);
-	uint64_t ticks_tout = domain->ticks_per_ms * LL_TIMER_PERIOD / 1000;
-	uint64_t ticks_set;
-	uint64_t current = platform_timer_get_atomic(zephyr_domain->timer);
+	uint64_t current = platform_timer_get_atomic(zephyr_domain->ll_timer);
 	uint64_t earliest_next = current + 1 + ZEPHYR_SCHED_COST;
-	uint64_t ticks_req = domain->next_tick ? start + ticks_tout :
-		MAX(start, earliest_next);
-	int ret, core = cpu_get_id();
-	uint64_t ticks_delta;
+	uint64_t ticks_tout = domain->ticks_per_ms * LL_TIMER_PERIOD_US / 1000;
```

Comment:
Not following why we need the domain_set() if Zephyr is doing the rearming via periodic timer API ?

Reply:
Btw, where is the timer handler code that runs in IRQ context ?

Reply:
I thought about that, and I think we might be able to remove it. We would then just start the timer in `.domain_register()`. I kept it so far for (1) theoretical power savings for cases when the user registers a domain but sets the timer start time far into the future, (2) higher start time precision at least for the first task. If you don't think any of these reasons are important enough, I can try to remove `.domain_set()`

Reply:
> Btw, where is the timer handler code that runs in IRQ context ?

@lgirdwood `compare_isr()` in drivers/timer/cavs_timer.c in Zephyr

Reply:
> I thought about that, and I think we might be able to remove it. We would then just start the timer in `.domain_register()`. I kept it so far for (1) theoretical power savings for cases when the user registers a domain but sets the timer start time far into the future, (2) higher start time precision at least for the first task. If you don't think any of these reasons are important enough, I can try to remove `.domain_set()`

Lets look at how pipelines will actually schedule LL and DP work. This API will come out of this.